### ISSUE TITLE
zero gpu usage when train imagenet

### ISSUE THREAD

**Issue by zcyang** [pytorch/examples#164]
When I train alexnet with the following command, the gpu usage is 0. It seems it get stuck with the data loader, which is very unstable and sometimes takes a long time for each batch. The loss does not decrease either...

```
CUDA_VISIBLE_DEVICES=0 python main.py -a alexnet --lr 0.01 -j 16 /mnt/sdb/imagenet --print-freq 1
```

```
Epoch: [0][14/5005]	Time 0.154 (8.010)	Data 0.001 (7.720)	Loss 6.9074 (6.9080)	Prec@1 0.000 (0.026)	Prec@5 0.391 (0.391)
Epoch: [0][15/5005]	Time 0.154 (7.519)	Data 0.001 (7.237)	Loss 6.9078 (6.9080)	Prec@1 0.000 (0.024)	Prec@5 0.781 (0.415)
Epoch: [0][16/5005]	Time 67.619 (11.054)	Data 67.544 (10.785)	Loss 6.9076 (6.9079)	Prec@1 0.000 (0.023)	Prec@5 0.781 (0.437)
Epoch: [0][17/5005]	Time 0.160 (10.449)	Data 0.001 (10.186)	Loss 6.9074 (6.9079)	Prec@1 0.000 (0.022)	Prec@5 0.000 (0.412)
Epoch: [0][18/5005]	Time 3.062 (10.060)	Data 2.995 (9.807)	Loss 6.9090 (6.9080)	Prec@1 0.391 (0.041)	Prec@5 0.781 (0.432)
Epoch: [0][19/5005]	Time 0.154 (9.565)	Data 0.001 (9.317)	Loss 6.9084 (6.9080)	Prec@1 0.000 (0.039)	Prec@5 0.781 (0.449)
Epoch: [0][20/5005]	Time 15.261 (9.836)	Data 15.193 (9.597)	Loss 6.9074 (6.9080)	Prec@1 0.000 (0.037)	Prec@5 0.391 (0.446)
Epoch: [0][21/5005]	Time 0.154 (9.396)	Data 0.001 (9.161)	Loss 6.9087 (6.9080)	Prec@1 0.000 (0.036)	Prec@5 0.781 (0.462)
Epoch: [0][22/5005]	Time 0.154 (8.994)	Data 0.001 (8.762)	Loss 6.9079 (6.9080)	Prec@1 0.000 (0.034)	Prec@5 0.000 (0.442)
Epoch: [0][23/5005]	Time 0.154 (8.626)	Data 0.001 (8.397)	Loss 6.9083 (6.9080)	Prec@1 0.000 (0.033)	Prec@5 0.000 (0.423)
Epoch: [0][24/5005]	Time 0.154 (8.287)	Data 0.001 (8.061)	Loss 6.9085 (6.9080)	Prec@1 0.000 (0.031)	Prec@5 0.781 (0.438)
Epoch: [0][25/5005]	Time 0.154 (7.974)	Data 0.001 (7.751)	Loss 6.9057 (6.9079)	Prec@1 0.781 (0.060)	Prec@5 1.562 (0.481)
Epoch: [0][26/5005]	Time 0.155 (7.685)	Data 0.001 (7.464)	Loss 6.9066 (6.9079)	Prec@1 0.000 (0.058)	Prec@5 0.781 (0.492)
Epoch: [0][27/5005]	Time 5.921 (7.622)	Data 5.846 (7.406)	Loss 6.9075 (6.9079)	Prec@1 0.000 (0.056)	Prec@5 0.000 (0.474)
Epoch: [0][28/5005]	Time 2.969 (7.461)	Data 2.900 (7.251)	Loss 6.9081 (6.9079)	Prec@1 0.000 (0.054)	Prec@5 0.000 (0.458)
Epoch: [0][29/5005]	Time 1.661 (7.268)	Data 1.593 (7.062)	Loss 6.9079 (6.9079)	Prec@1 0.000 (0.052)	Prec@5 0.391 (0.456)
Epoch: [0][30/5005]	Time 0.154 (7.038)	Data 0.001 (6.835)	Loss 6.9064 (6.9078)	Prec@1 0.000 (0.050)	Prec@5 1.172 (0.479)
Epoch: [0][31/5005]	Time 0.154 (6.823)	Data 0.001 (6.621)	Loss 6.9079 (6.9078)	Prec@1 0.000 (0.049)	Prec@5 0.391 (0.476)
Epoch: [0][32/5005]	Time 75.456 (8.903)	Data 75.379 (8.705)	Loss 6.9086 (6.9079)	Prec@1 0.000 (0.047)	Prec@5 0.000 (0.462)
Epoch: [0][33/5005]	Time 3.523 (8.745)	Data 3.455 (8.550)	Loss 6.9077 (6.9079)	Prec@1 0.000 (0.046)	Prec@5 1.172 (0.483)
Epoch: [0][34/5005]	Time 0.154 (8.499)	Data 0.001 (8.306)	Loss 6.9080 (6.9079)	Prec@1 0.000 (0.045)	Prec@5 0.391 (0.480)
Epoch: [0][35/5005]	Time 0.154 (8.268)	Data 0.001 (8.075)	Loss 6.9081 (6.9079)	Prec@1 0.000 (0.043)	Prec@5 0.781 (0.488)
Epoch: [0][36/5005]	Time 0.153 (8.048)	Data 0.001 (7.857)	Loss 6.9081 (6.9079)	Prec@1 0.000 (0.042)	Prec@5 0.000 (0.475)
Epoch: [0][37/5005]	Time 0.154 (7.840)	Data 0.001 (7.650)	Loss 6.9086 (6.9079)	Prec@1 0.000 (0.041)	Prec@5 0.391 (0.473)
Epoch: [0][38/5005]	Time 0.154 (7.643)	Data 0.001 (7.454)	Loss 6.9073 (6.9079)	Prec@1 0.000 (0.040)	Prec@5 1.172 (0.491)
Epoch: [0][39/5005]	Time 0.156 (7.456)	Data 0.001 (7.268)	Loss 6.9078 (6.9079)	Prec@1 0.391 (0.049)	Prec@5 1.562 (0.518)
Epoch: [0][40/5005]	Time 1.101 (7.301)	Data 1.035 (7.116)	Loss 6.9086 (6.9079)	Prec@1 0.000 (0.048)	Prec@5 0.000 (0.505)
Epoch: [0][41/5005]	Time 0.154 (7.131)	Data 0.001 (6.946)	Loss 6.9081 (6.9079)	Prec@1 0.000 (0.047)	Prec@5 0.000 (0.493)
Epoch: [0][42/5005]	Time 0.154 (6.969)	Data 0.001 (6.785)	Loss 6.9087 (6.9079)	Prec@1 0.000 (0.045)	Prec@5 0.000 (0.481)
Epoch: [0][43/5005]	Time 6.075 (6.948)	Data 6.007 (6.767)	Loss 6.9082 (6.9079)	Prec@1 0.000 (0.044)	Prec@5 0.781 (0.488)
Epoch: [0][44/5005]	Time 10.960 (7.038)	Data 10.883 (6.859)	Loss 6.9096 (6.9080)	Prec@1 0.000 (0.043)	Prec@5 1.172 (0.503)
Epoch: [0][45/5005]	Time 0.159 (6.888)	Data 0.001 (6.710)	Loss 6.9081 (6.9080)	Prec@1 0.000 (0.042)	Prec@5 0.781 (0.510)
```

**Comment by soumith** [comment]
for alexnet, it takes close to 1000 mini-batches for loss to start going down.

**Comment by zcyang** [comment]
I have tested on multiple machines and the same case happens. I also tested the disk read and write speed, they are normal.

**Comment by zcyang** [comment]
Is there any update on this problem? Thanks!

**Comment by micklexqg** [comment]
change the print frequency bigger and wait for a longer time

**Comment by Lan1991Xu** [comment]
Is this work? I also meet the same problem. The GPU utilization is almost 0

**Comment by bgardiner** [comment]
I am seeing the same result.  Data loading dominates training time (similar to as shown in @zcyang's output above).  Is the [pre-processing step](https://github.com/pytorch/examples/blob/master/imagenet/main.py#L116-L126) the bottleneck -- my disk I/O does not appear to be close to max speed?  Is there a way to pipeline that work to be done before GPU training so training does not take prohibitively long?

**Comment by MrWanter** [comment]
Same problem, gpu 0 usage, only small amount of time has usage

**Comment by hongyi-zhang** [comment]
Your dataloaders are taking up the bulk of the running time. It's likely that you are not assigning enough CPUs for the workers. Try e.g. at least 32 CPUs for 8 GPUs and 8 workers.
